### PR TITLE
Add support for async setup and teardown #60

### PR DESCRIPTION
**Pull request by simoneb**
This PR closes issue #60 by adding support for async setups and teardowns.

### CODE REVIEW

**Comment by rprouse** [comment]
Nice to see you back @simoneb :smile: 

I've had a quick look and it looks good. I am just heading out, so will take a closer look when I get back.

**Comment by simoneb** [comment]
Thanks @rprouse, it's only for a short while though, I've had this issue assigned for too long and Charlie prompted me to do something about it, rightfully so.

File: src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs [hunk 1]
```diff
@@ -84,12 +89,32 @@ public void RunTearDown(TestExecutionContext context)
...
                     while (--index >= 0)
-                        Reflect.InvokeMethod(_tearDownMethods[index], _tearDownMethods[index].IsStatic ? null : context.TestObject);
+                    {
+                        var tearDownMethod = _tearDownMethods[index];
+#if NET_4_0 || NET_4_5
+                        if (AsyncInvocationRegion.IsAsyncOperation(tearDownMethod))
+                            RunAsyncMethod(tearDownMethod, context);
+                        else
+#endif
+                            RunNonAsyncMethod(tearDownMethod, context);
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs]: You could extract this logic into a single method called for both setup and teardown.

**Comment by CharliePoole** [comment]
Looks ready to merge AFAIC, either with or without the refactoring I suggested. At some future time, we may want to do a more global refactoring and put the facility for invoking any method, async or not, into Reflect.cs.

File: src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs (hunk 1 above)
Comment (simoneb) [inline src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs]: Good point, I'll do this

**Comment by simoneb** [comment]
> Looks ready to merge AFAIC, either with or without the refactoring I suggested.

Refactoring done

> At some future time, we may want to do a more global refactoring and put the facility for invoking any method, async or not, into Reflect.cs.

I agree, the conditional everywhere `AsyncInvocationRegion` is used doesn't look too good

**Comment by oznetmaster** [comment]
I thought that all source files were now being included in every build, and that exclusion was being handled solely by conditionals.  That is not the case in this PR.

**Comment by CharliePoole** [comment]
@oznetmaster You're correct. Since it's merged and is not a breaking change, we should clean up when we are making some other change to the framework tests.

**Comment by Nezz** [comment]
Does this work? It appears to me that it doesn't.
https://github.com/nunit/nunit/blob/772eb236bc50b11c573010aaefa44cc1064c383d/src/NUnitFramework/framework/Internal/Tests/TestSuite.cs#L240

**Comment by Nezz** [comment]
Works if changed to
```C#
(!method.ReturnType.Equals(typeof(void)) && !method.ReturnType.Equals(typeof(Task)))
```

**Comment by simoneb** [comment]
@Nezz good point, my bad. I've submitted #489 to fix this, thanks for the heads up

**Comment by adamralph** [comment]
Is this feature documented anywhere? I can't find anything showing how to use it.

**Comment by manfred-brands** [comment]
I can't seem to find it either.
Simply change the return type of the method to `Task` and if needed add the `async` keyword.

**Comment by stevenaw** [comment]
I see a mention of it in the attribute documentation here:
- https://docs.nunit.org/articles/nunit/writing-tests/attributes/setup.html#notes
- https://docs.nunit.org/articles/nunit/writing-tests/attributes/teardown.html#notes

But it looks like in many places it isn't explicitly called out and the sample code is all non-async (ex: https://docs.nunit.org/articles/nunit/writing-tests/setup-teardown/)

**Pull request merged**